From c5fc6f31cdcdf296fd4b493970e89e3ccde3e586 Mon Sep 17 00:00:00 2001
From: "Brian C. Lane" <bcl@redhat.com>
Date: Fri, 20 Feb 2015 16:28:34 -0800
Subject: [PATCH] libparted: Use common function to calculate PTE sectors
 (#1180683)

Use _ptes_sectors in _parse_header's calculation to determine if the
disk has been grown.

Related: rhbz#1180683
---
 libparted/labels/gpt.c | 17 ++---------------
 1 file changed, 2 insertions(+), 15 deletions(-)

diff --git a/libparted/labels/gpt.c b/libparted/labels/gpt.c
index 3cf7385..6e33dc3 100644
--- a/libparted/labels/gpt.c
+++ b/libparted/labels/gpt.c
@@ -717,7 +717,7 @@ _parse_header (PedDisk *disk, const GuidPartitionTableHeader_t *gpt,
   GPTDiskData *gpt_disk_data = disk->disk_specific;
   PedSector first_usable;
   PedSector last_usable;
-  PedSector last_usable_if_grown, last_usable_min_default;
+  PedSector last_usable_if_grown;
   static int asked_already;
 
 #ifndef DISCOVER_ONLY
@@ -745,19 +745,7 @@ _parse_header (PedDisk *disk, const GuidPartitionTableHeader_t *gpt,
      space or continue with the current usable area.  Only ask once per
      parted invocation. */
 
-  last_usable_if_grown
-    = (disk->dev->length - 2 -
-       ((PedSector) (PED_LE32_TO_CPU (gpt->NumberOfPartitionEntries)) *
-        (PedSector) (PED_LE32_TO_CPU (gpt->SizeOfPartitionEntry)) /
-        disk->dev->sector_size));
-
-  last_usable_min_default = disk->dev->length - 2 -
-    GPT_DEFAULT_PARTITION_ENTRY_ARRAY_SIZE / disk->dev->sector_size;
-
-  if (last_usable_if_grown > last_usable_min_default)
-    {
-      last_usable_if_grown = last_usable_min_default;
-    }
+  last_usable_if_grown = disk->dev->length - 2 - _ptes_sectors(disk, gpt);
 
   PED_ASSERT (last_usable > first_usable, return 0);
   PED_ASSERT (last_usable <= disk->dev->length, return 0);
@@ -785,7 +773,6 @@ _parse_header (PedDisk *disk, const GuidPartitionTableHeader_t *gpt,
           ptt_clear_sectors (disk->dev,
                              gpt_disk_data->AlternateLBA, 1);
           gpt_disk_data->AlternateLBA = disk->dev->length - 1;
-          last_usable = last_usable_if_grown;
           *update_needed = 1;
         }
       else if (q != PED_EXCEPTION_UNHANDLED)
-- 
2.1.0

